### ISSUE TITLE
Reflexion on the french SAPIN LAW application

### ISSUE THREAD

**Issue by flotho** [OCA/l10n-france#105]
Dear Community, 

I would like to start a thread regarding the [SAPIN french law](https://www.legifrance.gouv.fr/affichTexteArticle.do?cidTexte=JORFTEXT000031732865&idArticle=LEGIARTI000031734740&dateTexte=20151230). From now the french administration is not really explicit about the ways to certify the Odoo solution. Does anyone has some legal info?

Here are some of the resources I found :
* http://revuefiduciaire.grouperf.com/article/3661/hb/20160919173548859.html
* https://www.april.org/point-detape-sur-la-nouvelle-reglementation-des-logiciels-de-gestion-et-de-comptabilite

Some part of the law seems to be easy to certify : Hosting / Backup / Recovery are easy for the community partners to be certified.

Regarding the durability of the datas I think we have a problem with the POS. From now the POS has been designed to be working without network and all the datas are stored inside the browser database. This point could be an issue if you consider how easy it is to get the datas from the internal database (and what about the debug mode allowing to flush the orders!!!)
I have some little ideas on those points :
* Find a way to block some inappropriate operation from the POS (like in debug mode)
* Find a way to store every transaction in the POSBox from the POS
* Find a way to protect the browser database with encryption
* Find a way to deliver browser application with options preventing the user from acceding to the internal database.

None of those solutions looks enough for me(everything in the client part could be changed by an experienced user/ ethic hacker).
 
Odoo seems to have started a reflexion on this : https://github.com/odoo-dev/odoo/commits/9.0-l10n_fr-certification-lpe . It looks like Odoo is considering that only the account_cancel module could be a problem.

Some partners have started a reflexion, (BTW thanks to Sébastien Morelle) : https://anybox.fr/blog/logiciels-de-caisse-certifies

Maybe we could start a thread here https://odoo-community.org/groups/france-24

Any feedback would be appreciated.

Regards

**Comment by sisalp** [comment]
hello,
this is the most operational document I know about this: http://brochures.sisalp.fr/referentiel-certification-systemes-caisse.pdf

Some comments 

- The POS is obviously a big issue. In my opinion, it requires deep reconsideration. At first, just don't use it at all.
- The account_cancel is another problem, but not for this law : rewriting accounting journals has been forbidden in France for past 500 years, it is just a patch to come back on track with legal accounting.

Some people do register payments using Odoo, in particular they consider that they can give the invoice at the desk and register the payment by setting the "paid" status of the invoice. They are wrong, already today. They must conform to payment regulations now, independently of this law. In particular it is true for non-VAT business which is not concerned by the new rule.

The law now underlines these obligations and requires tools used to register payments to get certified. This is the only new point compared with today situation.

People who are correct today, I mean who register their payments on a cash register or a paper book can keep on, no change.
People who used Odoo to register their payments should immediatly register them outside Odoo. Buy a cah register of a receipt official book.
People who are using the POS have a problem. They can 
- get a certificate from someone. Someone can be Odoo if they move to last enterprise, provided that we trust in Odoo capability to certify the Enterprise edition. Or someone can be a third party like Anybox, but here also we must be sure Anybox does understand the certification process, which their paper doesn't show (I made the same mistakes before I studied it deeper).
- replace the POS by another certified solution, for exemple by interfacing a cash register to Odoo.

Hope this helps.

**Comment by sebastienbeau** [comment]
@flotho regarding the point of sale on our side we think about

- removing the "offline feature" for the pos synchronisation. Indeed 2017 is cheaper to paid 2 Internet connexion than doing complex development, also it's easy to use your smartphone with 3G/4G. Today I have spend some time to check if it's possible, I just did a quick dirty poc for removing the synchronization here: https://github.com/OCA/pos/pull/172
- adding an hash computation on server side like in the odoo module for account move https://github.com/odoo/odoo/pull/16935

With that solution we are sure that no POS can be lost during the synchronization as the ticket will be generate only after the synchronization (and maybe we can generate it on server side so you can reprint a ticket from the backoffice easily).

We can also print the first letter/digit of the hash on the ticket, so every ticket are a proof of the inalterability.
On pure accounting side we also work on module to edit invoice after the validation for the field that do not impact your accounting so we can drop account_cancel

**Comment by legalsylvain** [comment]
Hi @flotho, 

Thanks a lot for beginning this thread ! We could in a first time enumerate each point, and after create little workshops for each point. I see the following points.

1. **Developpers - Accounting Part** : Secure accounting modules. Blocking account_cancel module, etc...
2. **Developpers - PoS Part** : Secure PoS features. Basically, what you said.
3. **Admin - Hosting Part** : what we have to do regarding hosting.  (forbid admin access, etc...)
4. **Legal - Customer Part** : what kind of documents we have to give to the customers. A community module to generate pdf could be welcome.
5. **Legal - Self Hosting Part** : Most of us are selling Odoo for customers, but we have an instance for us. Is it self-certification allowed ? 

do you see other points ? Thanks.

CC
- Anybox : @smorele
- Camp2Camp : @fclementic2c, @jgrandguillaume
- Akretion : @sebastienbeau, @alexis-via
- Syleam : @alnslang
- Acsone : @sbidoul
- Agipme : @mayjo79

kind regards.

**Comment by qdp-odoo** [comment]
Please, let us know your conclusions: on our side, so far, we don't plan to do anything more than the current l10n_fr_certification module proposition (except a signed document by Fabien certifying Odoo)

best regards

**Comment by sisalp** [comment]
Quentin : do you mean Odoo Enterprise will not pass through certification process ?

**Comment by fmdl** [comment]
I think there are an other point : it the software fingerprint. 
In page 8 of http://brochures.sisalp.fr/referentiel-certification-systemes-caisse.pdf, the certifcation need  the software fingerprint. it is possible to calculate the fingerprint of odoo code and compare with the fingerprint save on Odoo website. (that like fiscal administration can know if 
But there are an issue with the external addons, every addons can access to every methode, I don't know if it is possible with python to protect the account methode. (or protect cr.execute)

Maybe use https://www.postgresql.org/docs/current/static/pgcrypto.html or http://docs.postgresql.fr/9.2/ssl-tcp.html to crypt database.

**Comment by alexis-via** [comment]
I propose we speak about this during the OCA code sprint in Barcelona https://odoo-community.org/event/barcelona-code-sprint-2017-05-29-2017-05-31-58/register It will be a good opportunity to exchange our ideas and update the status of the work on this. At Akretion, we already started some devs on this topic and we will continue (maybe during the code sprint too).

**Comment by alexis-via** [comment]
For those who missed the info : https://www.lesechos.fr/thema/030387673950-revirement-bienvenu-en-matiere-de-logiciels-certifies-2094788.php
The law will be changed and should only target POS software (not accounting software).

**Comment by legalsylvain** [comment]
I just made a PR to manage certification for PoS. (and for account, even if it does'nt seems necessary anymore, given the last @alexis-via remark).
Thanks for your review.

https://github.com/OCA/l10n-france/pull/108

**Comment by flotho** [comment]
Thanks to the community for the tips http://proxy-pubminefi.diffusion.finances.gouv.fr/pub/document/18/22503.pdf

**Comment by Auneor** [comment]
Hi everyone
How about centralizing/synchronizing the dev/discussions between all partner? I think the complexity of the law/devs needed make it difficult to have everything in a single thread like here. 
So some devs have been done in order to backport the odoo module l10n_fr_certification and make the pos compliant for v8 https://github.com/OCA/l10n-france/pull/108
I think there is still some problems, with the approach that odoo propose; indeed, nothing is proposed to close an accounting period, see the point 110 in http://bofip.impots.gouv.fr/bofip/10691-PGP
Another problem, is with a test/demo database, that must modify reports and pos tickets to include "TEST" or "FACTICE" (point 150 from the previous URL)
I didn't see anything done in l10_fr_certification for v9/10 to secure/hash PoS operation like payments, bill emited, I think it's critical and needed to comply with the law (point 130 and 140)
We made a github repository here https://github.com/MohammedAuneor/lf2016 (because we didn't find this thread here when we decided to do it) but we are open to move the discussion somewhere else, here, or on other support. 
The approach I was thinking was more to use a global table with hashes for all the application, and to choose more carrefully what is hashed and stored in a secure/inalterable way. The only problem I see is with the point 220, that says:  _Elle doit prévoir un dispositif technique garantissant l'intégrité dans le temps des archives produites et _leur conformité aux données initiales de règlement à partir desquelles elles sont créées.__ 
Regards

**Comment by legalsylvain** [comment]
Hi @MohammedAuneor,

Welcome in the community.

> How about centralizing/synchronizing the dev/discussions between all partner? I think the complexity of the law/devs needed make it difficult to have everything in a single thread like here.

Well, this thread has a avantage to join various people of the OCA community. Let's go continuing with this, and if we have to create later work group, it will be more easy.

> I think there is still some problems, with the approach that odoo propose; indeed, nothing is proposed to close an accounting period, see the point 110 in http://bofip.impots.gouv.fr/bofip/10691-PGP

Accouting is now out of the scope of the law, just Point of sale. See @alexis-via comment. (https://github.com/OCA/l10n-france/issues/105#issuecomment-309730886).
Anyway, this part should be out of the l10n-france thread, because many countries require to close accounting period, and it's a global problem.
And even if, it is managed by the module l10n_fr_certification_account, (#108). As all the entries are ordered, it is trivial to check that an 2016 entry has been written in 2017.

> Another problem, is with a test/demo database, that must modify reports and pos tickets to include "TEST" or "FACTICE" (point 150 from the previous URL)

Good idea ! for me, it's optional, because not all configurations provide test / demo servers.
But overloading all report will be a mess, and will not be very secure, as it's possible for end user to edit custom report.
What about a more light approach ? As company name is set on all the report and bill, It will be more easy to just rename your company, or to have a little module, depending on xx_environment OCA / server-tools module that overload name_get function of res.company, to add (DEMO) at the end of the name.

> I didn't see anything done in l10_fr_certification for v9/10 to secure/hash PoS operation like payments, bill emited, I think it's critical and needed to comply with the law (point 130 and 140)

Indeed, we have to port l10n_fr_certification_pos module. Feel free to review  #108. It should be cool to have a stable 8 version, before porting it. and it will not need a lot of work.( some JS work)

> The only problem I see is with the point 220, that says: Elle doit prévoir un dispositif technique garantissant l'intégrité dans le temps des archives produites et leur conformité aux données initiales de règlement à partir desquelles elles sont créées.

I don't understand that point. This is a matter of adminsys people, to grant backup. What do you propose ? 

Thanks for your comment.
kind regards.

**Comment by sisalp** [comment]
2017-07-21 16:46 GMT+02:00 Sylvain LE GAL <notifications@github.com>:


> The only problem I see is with the point 220, that says: Elle doit prévoir
> un dispositif technique garantissant l'intégrité dans le temps des archives
> produites et leur conformité aux données initiales de règlement à partir
> desquelles elles sont créées.
>
> I don't understand that point. This is a matter of adminsys people, to
> grant backup. What do you propose ?
>

I'm afraid
1_ this point is the main one and we must all understand it.
2_ no, it is not a sys-admin aspect, but THE feature any POS software MUST
conform to. All others are at best "nice-to-have" only.

**Comment by legalsylvain** [comment]
Hi @sisalp. Please, be more precise.

1. What do you propose ?
2. Whad did you plan to do for your enterprise ? 

regards.

**Comment by sisalp** [comment]
2017-07-21 19:29 GMT+02:00 Sylvain LE GAL <notifications@github.com>:
>
> Hi @sisalp. Please, be more precise.
>
> What do you propose ?


This law about a set of precise requirements, if you want to build a
solution which conforms, you have to pass the qualification tests. The
requirement Auneor talks about is key.
AFAIU, some may make an error when they think they may issue
attestations on a software which definitly doesn't pass most of the
tests or with even no plan at all.
I follow this thread trying to understand if the community will
address the questions:
- is it feasable first, then maintainable ?
- are there a migration path and service offreings for existing customers.

>
> Whad did you plan to do for your enterprise ?

For my enterprise ? I don't use Odoo.
For my hosted customers who use the POS (few), either they ask their
integrator, or I redirect them to people who propose solutions (few
also) and they make their own choice. I'm not sure going down to more
details would help in this discussion.

**Comment by jcchoquet** [comment]
For those who missed the info : https://www.economie.gouv.fr/files/files/directions_services/dgfip/controle_fiscal/actualites_reponses/logiciels_de_caisse.pdf
Accouting is in of the scope of the law, for people not subject to VAT.

**Comment by fgi-odoo** [comment]
We are going to upgrade l10n_fr_certification for POS module in order to be compliant with the new legislation that we read carefully. This new module will be released asap and will be available for versions 9, 10 & 11. 

Here are the functionalities we plan to cover:
- no way to cancel POS orders and journal entries
- no way to cancel or modify the balance of payment journal statements (access from POS sessions)
- hash computation on POS orders and journal entries
- interface to check the inalterability of POS orders and journal entries
- Computation of payment period's grand total (daily, weekly, monthly periods) > new object
- download of the mandatory certificate (only available for Odoo Enterprise & Online)

Is there anything else we should take into account? Have you already developed new modules covering those aspects? Your feedback will be much appreciated. Thanks!

**Comment by sisalp** [comment]
2017-10-16 16:39 GMT+02:00 fgi-odoo <notifications@github.com>:

>
> Is there anything else we should take into account? Have you already developed new modules covering those aspects? Your feedback will be much appreciated. Thanks!

What is the level of coverage of theses tests  :
http://brochures.sisalp.fr/referentiel-certification-systemes-caisse.pdf
?
Which don't conform ?

**Comment by legalsylvain** [comment]
Hi @fgi-odoo. Thanks a lot to ask to the community. Very appreciated. Are you responsible of Point Of Sale in Odoo ? 

The point are pretty clear for me, except two : 

- cumulative sessions for POS (daily, weekly, monthly) + computation of session's grand total

Are talking about the possiblity to generate pdf with daily / weekly / etc.. information ? Not sure to understand.

- download of the mandatory certificate (only available for Odoo Enterprise & Online)

Are you talking about the possibility to generate via Odoo the certificate. (pdf too).

thanks for your precision.

kind regards.

**Comment by legalsylvain** [comment]
@sisalp : your document is outdated. Don't it ? (décembre 2016)

After the election of Macron, see  @jcchoquet remarks. If you find a up-to-date document, please share to the community.

regards.

**Comment by jcchoquet** [comment]
Hi @fgi-odoo,
if you see this link : https://www.economie.gouv.fr/files/files/directions_services/dgfip/controle_fiscal/actualites_reponses/logiciels_de_caisse.pdf
it is not only the POS of concerned, but also the customer payments in Odoo...

**Comment by sisalp** [comment]
2017-10-16 17:42 GMT+02:00 Sylvain LE GAL <notifications@github.com>:
>
> @sisalp : your document is outdated. Don't it ? (décembre 2016)


I didn't see any evolution about how a device can be tested in the
law. All discussions I saw were about environment (who, how, why...),
not about compliance.
>
> After the election of Macron, see @jcchoquet remarks. If you find a up-to-date document, please share to the community.

I don't know about these remarks. Did they change the requirements ?
Any pointer ?

>
> regards.

**Comment by sisalp** [comment]
Answering to my-self: I guess you refer to a comment in this list. Yes
of course, I followed this carefully. The document I pointed is up to
date from my standpoint.

2017-10-16 18:03 GMT+02:00 Dominique Chabord <dominique.chabord@sisalp.org>:
> 2017-10-16 17:42 GMT+02:00 Sylvain LE GAL <notifications@github.com>:
>>
>> @sisalp : your document is outdated. Don't it ? (décembre 2016)
>
>
> I didn't see any evolution about how a device can be tested in the
> law. All discussions I saw were about environment (who, how, why...),
> not about compliance.
>>
>> After the election of Macron, see @jcchoquet remarks. If you find a up-to-date document, please share to the community.
>
> I don't know about these remarks. Did they change the requirements ?
> Any pointer ?
>
>>
>> regards.



-- 
Dominique Chabord - SISalp
Logiciel libre pour l'entreprise Tryton et open-source Odoo, OpenERP
Les Millières 74230 Serraval - France
tel(repondeur) +33(0)950274960 fax +33(0)955274960 mob +33(0)622616438
http://sisalp.fr
http://boutique.sisalp.fr
https://twitter.com/SISalp l'actualité de vos services en temps réel.

**Comment by jcchoquet** [comment]
when I have contacted the DGFIP for this law, i have this response :
![image](https://user-images.githubusercontent.com/29275440/31623094-a7cc0542-b29e-11e7-96ac-d182f13d19df.png)

**Comment by legalsylvain** [comment]
Hi @jcchoquet thanks a lot for your link.

regards.

**Comment by fgi-odoo** [comment]
@legalsylvain yes I'm the product owner for all the sales apps (sales, pos, ecommerce, etc.). 
- Cumulative sessions: This refers to topics 28, 29 & 30 of https://www.economie.gouv.fr/files/files/directions_services/dgfip/controle_fiscal/actualites_reponses/logiciels_de_caisse.pdf. We need to store the total of daily, weekly & monthly sessions and include a sum check in the inalterability tests. Therefore we think about a new business object, atop the journal entries for payments.
- Certificate: yes it is to download it in PDF.

@jcchoquet: for now balance of POS payment journals can be edited from the Accounting menu since statements stay in draft as long as you don't close the session. So the idea is to prevent the editing of such draft statements if they relate to POS journals. Therefore the only way to make corrections is through the frontend with plus and minus operations, as stated in the new regulation.

**Comment by jcchoquet** [comment]
@fgi-odoo : the new regulation is not only for cash but for all methods of payment (see topic 11 of link)
for me, Accounting is in of the scope for the sales "BtoC".

**Comment by fgi-odoo** [comment]
@jcchoquet Indeed! So we will stick to the initial plan and prevent users from cancelling any journal entry, including confirmed account payments.

**Comment by jcchoquet** [comment]
the certification will be made also for the community version ?

**Comment by fgi-odoo** [comment]
No. To be certified the user will be requested to download the certificate. And this certificate will be only delivered to Enterprise customers.

**Comment by jcchoquet** [comment]
Otto Runarsson said me that the community version was certified...
the accounting is in the community version, what do we will say to the customers which have the community version?

**Comment by fgi-odoo** [comment]
They will have to upgrade to Odoo Enterprise.

**Comment by jcchoquet** [comment]
for the small companies or the companies with lot of specific developpment it is not easy! I think this companies will must leave Odoo, the cost is too important to migrate...

**Comment by tkFontaine** [comment]
Hi,
[That](https://www.odoo.com/apps/modules/9.0/l10n_fr_certification/ ) module isn't enought to validate certification ?
@fgi-odoo What do you mean "Enterprise customers" ? If I'm a partner, will I get that certificate ?

**Comment by jcchoquet** [comment]
@fgi-odoo, see the topic 32 of link, the certificate of the community version is possible (as for the entreprise version), if the integrator modify the accounting module it's him who is responsable...

**Comment by tkFontaine** [comment]
@jcchoquet Who's Otto Runarsson ?

**Comment by jcchoquet** [comment]
a business consultant Odoo...

**Comment by jcchoquet** [comment]
last news :
As of January 1st 2018, French companies registering payments from their customers using a cash registering and/or an accounting system will be required to satisfy to certain conditions in terms of inalterability, security, storage and archiving of data (CGI art. 286, I. 3°bis). 

Very soon, Odoo will launch a new module to install in your system in order to comply with those new requirements. Through it, Odoo Online and Enterprise users will be able to download the mandatory compliance certificate to deliver to the tax authorities in the event of a fiscal audit.
This certificate will not be delivered to users of Odoo Community.

**Comment by sisalp** [comment]
thank you
any pointer to a source ?

2017-10-18 16:37 GMT+02:00 Oxilia-info <notifications@github.com>:
> last news :
> As of January 1st 2018, French companies registering payments from their
> customers using a cash registering and/or an accounting system will be
> required to satisfy to certain conditions in terms of inalterability,
> security, storage and archiving of data (CGI art. 286, I. 3°bis).
>
> Very soon, Odoo will launch a new module to install in your system in order
> to comply with those new requirements. Through it, Odoo Online and
> Enterprise users will be able to download the mandatory compliance
> certificate to deliver to the tax authorities in the event of a fiscal
> audit.
> This certificate will not be delivered to users of Odoo Community.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.



-- 
Dominique Chabord - SISalp
Logiciel libre pour l'entreprise Tryton et open-source Odoo, OpenERP
Les Millières 74230 Serraval - France
tel(repondeur) +33(0)950274960 fax +33(0)955274960 mob +33(0)622616438
http://sisalp.fr
http://boutique.sisalp.fr
https://twitter.com/SISalp l'actualité de vos services en temps réel.

**Comment by jcchoquet** [comment]
![image](https://user-images.githubusercontent.com/29275440/31725595-05709cda-b425-11e7-9986-db4bef13a4c6.png)

**Comment by jcchoquet** [comment]
the source is an email of Odoo.

**Comment by sisalp** [comment]
Thank you.

For non-french readers, the letter from Odoo says that Odoo-Enterprise
and Odoo-Saas will be attested by Odoo-SA. Traçability of code will be
insured by an additional module. No mention of any restriction of any
kind to keep this attestation valid.

So if you trust in Odoo, you can expect your system to actually
conform to the rules and you will be protected from lawsuits. We can
expect the format of the attestation will be the one demanded by the
admnistration.

This fits with previous comitments made byOdoo a year ago.

**Comment by legalsylvain** [comment]
Hi @fgi-odoo, 
Thanks for your answers. I don't understand the following point : 

> No. To be certified the user will be requested to download the certificate. And this certificate will be only delivered to Enterprise customers.

I don't understand why ? This is the same code for point of sale CE and EE. (loyalty module excepted, that has nothing to do with the certificate). So, if you can provide a certificate for EE, you can do it too for CE. Or did I miss something ?

thanks for the clarification.

**Comment by alexis-via** [comment]
@legalsylvain @fgi-odoo The certification is an engagement that the issuer of the certificate takes for his customer on the behavior of the software. The issuer of the certificate can get a fine the software is not compliant. So it is obvious that no company can do that for free.

But the law says that an integrator can also issue the certificate for this customer, cf 
http://bofip.impots.gouv.fr/bofip/10691-PGP.html?identifiant=BOI-TVA-DECLA-30-10-30-20160803
section 310 second entry. So, if a company runs Odoo CE, it can ask its integrator to issue a certificate. Then the integrator who issue the certificate will support the associated responsability and risks for that customer.

**Comment by jcchoquet** [comment]
@alexis-via, I am Ok with your second point.
But Odoo could distribute the famous new module that allows certification for the community version (for a fee par exemple) and that it is the integrator who certifies the solution to his client, no ?

**Comment by legalsylvain** [comment]
@alexis-via :

- There is a difference between doing it for free and force users to switch to the EE Edition. Do you see the nuance ?
- I just say that once the certification work is done. (code compliance, etc.) It doesn't need extra work to certificate the CE edition as it is the same code.

So I think that do not provide certification of Odoo CE is a political choice of Odoo SA, which will harm open source side of Odoo, (at least in France). And saying that it is enough to contract with an integrator is not pro - open source comment.

regards.

**Comment by fgi-odoo** [comment]
The module will be available on community version of course. As long as Odoo provides the certificate, this will be only for Enterprise users. However we plan to get certified by an official certifier early 2018. If this really happens the certifier would distribute the certificate to any Odoo user.

**Comment by jcchoquet** [comment]
@fgi-odoo , it's really official ?

**Comment by tkFontaine** [comment]
@fgi-odoo Thank you for your feedback. 
Could you explain what do you mean by "_the certifier would distribute the certificate to any Odoo user_" ?
Officialy, there are only 2 certifier (LNE & INFOCERT).

**Comment by fgi-odoo** [comment]
It's not official. By law we are not allowed to communicate more details about that unfortunately. And for sure this would not happen before 2018. The implications of that still need to be discussed. 

Our goal for now is to provide the community asap with a module that makes Odoo compliant with the functional requirements of the new legislation. When it comes to the certificate, Odoo will deliver it to Enterprise users. It's an important commitment of Odoo legally speaking so we can't provide it to anyone without control. As said earlier, any integrator can deliver its own certificate to its Community customers.

We'd appreciate your feedback about the specifications listed here above. More especially about the computation of period's cumulative grand total and perpetual total for daily, weekly and monthly payment sessions (see topic 28 in FAQ). It seems to us that we need a new "session" object above the payment journal entries. But POS payments are only posted once the POS session closed, and there is no time constraint on closing POS sessions. What do you think? How would you do that?

Thanks!

**Comment by jcchoquet** [comment]
Not before 2018 ? but the law must be applied at 1st january.
You think the module will be available before 2018 ?
thanks for the community to provide a module!! it's a good news.
For me, the payments must be saved immediately, for the POS or not (payment on invoice). See topic 19 and 20 in FAQ, if a payment must be modified, the modification must be trace.

**Comment by fgi-odoo** [comment]
@jcchoquet Yes as we said the module will be released before, as well as the access to the Odoo certificate. 

The POS payments are saved in draft payment statements. The idea is to lock such statements to not edit them. When the POS session is closed by the user, those statements are confirmed and posted. I think it's fine to compute cumulative payment amounts on confirmed journal entries only, and not on unconfirmed payment POS statements. Otherwise it  would make the logic complex. What's your opinion on that?

Regarding the payments registered on invoices, journal entries are posted straight away. So our solution would work fine as well.

**Comment by jcchoquet** [comment]
ok, great.
for the POS, the problem is if the session is not closed ?
I think the POS payments must be confirmed and posted immediately, because a unconfirmed statement can be cancelled ? see the topic 20.

Yes to compute cumulative payment amounts on confirmed journal.

**Comment by fgi-odoo** [comment]
It would mean a payment statement per transaction (rather than per session), and important changes in the code to do so. I don't like it. However making such POS statements non editable (to not cancel them) is easy to do. Users would still have to close POS sessions everyday, so that new payments would be included in daily cumulative totals. It seems to me this condition could be explicitly written in the system so that the users know how to be compliant.

**Comment by sisalp** [comment]
2017-10-26 16:25 GMT+02:00 fgi-odoo <notifications@github.com>:
> It would mean a payment statement per transaction (rather than per session),

For me, it is the most important point this law is about.
Everytime a ticket is paid, the controller wants to see a new and
definitive record and updated daily totals.

**Comment by jcchoquet** [comment]
@sisalp, I'am agree with you.
for me the POS payments must be confirmed and posted immediately...

**Comment by fgi-odoo** [comment]
Nothing requires to post them, but to lock them. We can warn in the report for the controller that sessions opened before today are not closed yet. Same on the POS session we can add a blocking warning if a user tries to launch a session once the opening day passed. This is much cleaner than hacking the way payment statements are posted. It seems to us this would fit the new requirements. If you think it wouldn't, could you please elaborate? Thanks!

**Comment by sisalp** [comment]
2017-10-26 17:36 GMT+02:00 fgi-odoo <notifications@github.com>:
> Nothing requires to post them,

Correct, posting is not the requirement

> but to lock them.

Wrong, the requirement is to record every payment with timestamped
data in its definitive form.

> We can warn in the report
> for the controller that sessions opened before today are not closed yet.
> Same on the POS session we can add a blocking warning if a user tries to
> launch a session once the opening day passed.

The warning 'hey, something is ongoing which is not fully recorded" is
equivalent to "no, I don't conform because I cannot show data of the
latest payment and its irreversible impact of totals".

It is the main test of the document I pointed already : Controller
makes a new payment and editor shows that it is recorded definitively
and that daily total is updated.
Any attestation just claims this test succeeds under any conditions.
In practice, any control will start this way: the controller buys
something, pays and gets the ticket. Now, he says "Hi, I'm the
controller, please show me where my payment data are recordered"

Nevertheless, I may miss the point in your question. Good news is that
if the payment is definitively recorded, nobody cares when/how it is
copied to the accounting.


>This is much cleaner than
> hacking the way payment statements are posted. It seems to us this would fit
> the new requirements. If you think it wouldn't, could you please elaborate?
> Thanks!

Clean is not a requirement.
Easy-to-control is a requirement.

hope it helps.

**Comment by fgi-odoo** [comment]
@sisalp "Controller
makes a new payment and editor shows that it is recorded definitively
and that daily total is updated." > I don't read that in the rules. Could you give the official text you refer to?

Here is the only paragraph I found about that:

170
Les systèmes de caisse doivent, de plus, prévoir obligatoirement une clôture journalière et une clôture mensuelle. Pour chaque clôture - journalière, mensuelle et annuelle (ou par exercice) - des données cumulatives et récapitulatives, intègres et inaltérables, doivent être calculées par le système de caisse, comme le cumul du grand total de la période et le total perpétuel pour la période comptable.

So I don't see anything missing in our specs. Cumulative totals are for each closing, not for each transaction. Let me know if I missed something.

**Comment by sisalp** [comment]
2017-10-27 9:00 GMT+02:00 fgi-odoo <notifications@github.com>:
>
> @sisalp "Controller
> makes a new payment and editor shows that it is recorded definitively
> and that daily total is updated." > I don't read that in the rules. Could you give the official text you refer to?


you may be correct, I maybe over specify.
In the Condition 15 of the LNE certification process, the software
must pass this test :
Clôture journalière :
-
s'authentifier sur le dispositif avec le rôle desti
né à l'utilisateur du dispositif ;
-
réaliser des opérations d'enregistrements sur cette
période journalière ;
-
accéder aux données d'enregistrement ;
-
vérifier la bonne génération d'une clôture journal
ière.

Which means that there is a closing action before total is updated.


>
> Here is the only paragraph I found about that:
>
> 170
> Les systèmes de caisse doivent, de plus, prévoir obligatoirement une clôture journalière et une clôture mensuelle. Pour chaque clôture - journalière, mensuelle et annuelle (ou par exercice) - des données cumulatives et récapitulatives, intègres et inaltérables, doivent être calculées par le système de caisse, comme le cumul du grand total de la période et le total perpétuel pour la période comptable.
>
> So I don't see anything missing in our specs. Cumulative totals are for each closing, not for each transaction. Let me know if I missed something.
>

No you are right. Reading again the test procedure, totals are
verified at closing time.

**Comment by fgi-odoo** [comment]
Eventually we suggest to compute cumulative & perpetual amounts on a new "Closing" object, based on sales entries (journal type = Sale) and not on the payment entries as I initially said. Indeed the rule requires it on revenue, not on payments. So it seems to us that open invoices should be taken into account as well.

**Comment by tkFontaine** [comment]
@fgi-odoo About certificate, I'm not sure to understand, you said first 

> And for sure this would not happen before 2018.

And then 

> Yes as we said the module will be released before, as well as the access to the Odoo certificate.

So will it be available before 1st January 2018 ?

Then, is it possible for you to explain what will you do for POS sales ? You can offline sell in POS, so the sales  is store in your browser cache ==> User can access to it and modify sales. It should not be able to.

**Comment by fgi-odoo** [comment]
@tkFontaine What we do for POS is explained here: https://github.com/OCA/l10n-france/issues/105#issuecomment-336907598

Yes our module will be available soon, for sure before January 1st. We are currently finishing the development. We will share the pull request asap so that you can take a look and provide your feedback.

One idea at the beginning was to prevent users from using the POS offline, so that they were not able to modify the cache entries. Not sure we have to go that far. It might annoy users in case of server disconnection. The law clearly stipulates that a "man of the art" will always be able to crack the system. I'd like to hear the opinion of the other fellows.

**Comment by sisalp** [comment]
2017-11-02 11:04 GMT+01:00 fgi-odoo <notifications@github.com>:

> One idea at the beginning was to prevent users from using the POS offline,
> so that they were not able to modify the cache entries. Not sure we have to
> go that far.


What is sure is that as soon as a payment is done by the customer,
payment information must be definitively recorded. (see recent posts,
and indeed, I was wrong about the totals). If you can do this with
off-line POS, it is OK. You don't seem to do it.

**Comment by jcchoquet** [comment]
@sisalp I'm Ok with you, a payment is done by the customer, payment information must be definitively recorded, for the POS or a simple invoice.

**Comment by sisalp** [comment]
2017-11-02 14:09 GMT+01:00 Oxilia-info <notifications@github.com>:
> @sisalp I'm Ok with you, a payment is done by the customer, payment
> information must be definitively recorded, for the POS or a simple invoice.


sorry,
in order to avoid over-specification, I did not say that.,
It seems that if direct payments are all recorded on the POS, you do
not need anyhing specific on onvoice.

**Comment by jcchoquet** [comment]
No no, I say just that Accounting is also in of the scope for the sales "BtoC", as the POS..

**Comment by sisalp** [comment]
2017-11-02 15:07 GMT+01:00 Oxilia-info <notifications@github.com>:
>
> No no, I say just that Accounting is also in of the scope for the sales "BtoC", as the POS..
>
I think it is important to make it clear that we do not agree on this point.

**Comment by jcchoquet** [comment]
for you, Accounting is also in of the scope, ok?

**Comment by sisalp** [comment]
2017-11-02 15:49 GMT+01:00 Oxilia-info <notifications@github.com>:
>
> for you, Accounting is also in of the scope, ok?
>
>
For me,
Accounting has its own rules to follow
Accounting is not concerned by Sapin Law
One can use Odoo POS and not use Accounting.

Nevertheless, I don't know how POS is to be secured and missed some
discussions around Odoo. If securing the whole Odoo is the strategy to
get the POS safe, it may be a solution, but it is not the requirement
from the Sapin's law.
When I installed the POS alone, I got 34 modules installed, which may
indicate that the POS function is deeply nested in Odoo.

**Comment by jcchoquet** [comment]
accounting no, but invoicing and payment registrer are concerned..
this is the responses of DGFIP (in french): 
![image](https://user-images.githubusercontent.com/29275440/32340305-7e0c8984-bffa-11e7-8198-9ccfebd69cd0.png)

**Comment by sisalp** [comment]
2017-11-02 18:20 GMT+01:00 Oxilia-info <notifications@github.com>:
> accounting no, but invoicing and payment registrer are concerned..
> this is the responses of DGFIP (in french):
>

Please don't reopen the box :-(
These points have been discussed thousands of times since two years.
Just consider we disagree about what the requirement is for Odoo.
Next step is to check how Odoo will conform to certifications tests,
for example those in the test procedure I mentioned. This is what the
discussion is about today.

**Comment by jcchoquet** [comment]
Why ?
these points don't have been discussed since 2 years, because the DGFIP FAQ with the new rules is late July 2017.
For that Odoo will conform to certification, all th rules must be respected!

**Comment by sisalp** [comment]
2017-11-02 19:15 GMT+01:00 Oxilia-info <notifications@github.com>:
>
> Why ?
> these points don't have been discussed since 2 years, because the DGFIP FAQ with the new rules is late July 2017.
> For that Odoo will conform to certification, all th rules must be respected!

The document you extract is also obsolet. The law has been quite
deeply amended in september.
I just bet that Odoo know what they have to do while I don't see
clearly where they plan to go regarding conformance to official test
procedures. For sure, it is too late to discuss what this law is
about, but just imho.

**Comment by jcchoquet** [comment]
the responses of DGFIP are late October 2017.
Odoo takes responsibility with this certification, so I hope they know what to do... case to follow...

**Comment by tkFontaine** [comment]
@sisalp you said

> The law has been quite deeply amended in september

Where did you see that ? Is there a new document ?

**Comment by sisalp** [comment]
2017-11-03 4:30 GMT+01:00 tkFontaine <notifications@github.com>:
> @sisalp you said
>
> The law has been quite deeply amended in september
>
> Where did you see that ? Is there a new document ?
>


A final text is probably available because the 2018 law is now approved.
For the detail and a enlighted comment, see for example (in french):
https://philippe.scoffoni.net/loi-de-finance-2016-concerne-ou-pas/


Nevertheless, all 2017 evolutions were about "who is impacted" and
which functions of the software must conform. It also limited the
abusive interpretation the administration was trying to enforce.

We've always aknowledged that POS was the target. It was unclear
whether a backoffice software could be impacted. It is now clear that
only POS is concerned, and also that any direct payment has to be
recorded on a certified POS or cashbox.
Nevertheless, IIUC, Odoo seems to secure at backoffice level, maybe
because back and front are too intricated in Odoo, I don't know.
Securing the server is accepted by the law despite possible exploits.

There is no change I'm aware of about the conformance tests
themselves. If the Odoo POS passes the tests, its is OK, that's it.

**Comment by jcchoquet** [comment]
Odoo seems to secure at backoffice level, because the sales "BtoC" are in the scope of law, that's it.
Actually,The "private" taxable person is not subject to the invoicing obligation within the meaning of Article 289 of the CGI.
"(4) « 3° bis Si elle effectue des livraisons de biens et des prestations de services ne donnant pas lieu à facturation conformément à l’article 289 et enregistre ces opérations au moyen d’un logiciel ou d’un système de caisse, utiliser un logiciel ou un système satisfaisant à des conditions d’inaltérabilité, de sécurisation, de conservation et d’archivage des données en vue du contrôle de l’administration fiscale, attestées par un certificat délivré par un organisme accrédité dans les conditions prévues à l’article L. 433-4 du code de la consommation ou par une attestation individuelle de l’éditeur, conforme à un modèle fixé par l’administration ; »"

**Comment by jcchoquet** [comment]
@sisalp , if you have a reference that says the opposite, we are interested.

**Comment by sisalp** [comment]
Le 3 novembre 2017 à 10:52, Oxilia-info <notifications@github.com> a écrit :
>
> Odoo seems to secure at backoffice level, because the sales "BtoC" are in the scope of law, that's it.

Please don't make answers for them. It makes the question even more complicated.


>
> Actually,The "private" taxable person is not subject to the invoicing obligation within the meaning of Article 289 of the CGI.
> "(4) « 3° bis Si elle effectue des livraisons de biens et des prestations de services ne donnant pas lieu à facturation  »"
>

It is self-explaining for me, and I don't see how it relates to the discussion.
I'm afraid we have open this endless discussion again.
I'll try not to react anymore on reactions on my reactions...

**Comment by PatrickSUDOKEYS** [comment]
Hello chanel 
Sorry for my late arrival on this subject. At SUDOKEYS, we agree with the general position, only the POS is concerned. The initial law is modified by the  « Communiqué de presse n°22 du 15 juin 2017, Gérald Darmanin, ministre de l'Action et des comptes publics »
Stipulant :
"Face à l’inquiétude exprimée par les entreprises, notamment les plus petites d’entre elles, quant à la mise en œuvre au 1er janvier 2018 d’un dispositif de la loi de finances pour 2016 visant l’usage de logiciels de caisse, de comptabilité et de gestion certifiés, le Ministre de l’Action et des Comptes publics Gérald DARMANIN a décidé de le recentrer et de le simplifier. Seuls les logiciels et systèmes de caisse, principaux vecteurs des fraudes constatées à la TVA, seront ainsi concernés. Sans réduire son efficacité pour lutter contre les fraudes permises par l’apparition de logiciels permettant d'effacer des recettes enregistrées, la redéfinition du périmètre de l’obligation permet d’alléger la complexité induite, tant pour la mise en conformité initiale que pour le quotidien des entreprises. Cette modification fera l’objet de mesures législatives d’ici la fin d’année, pour une entrée en vigueur du dispositif comme prévu au 1er janvier 2018. Les entreprises qui n'auraient pas encore effectué cette mise en conformité de leur logiciel de caisse ont ainsi 6 mois pour y veiller. Le Ministre de l’Action et des Comptes publics demande à l’administration fiscale d’accompagner les entreprises dans la première année d’application des nouvelles règles."

It should be noted that this amending law is still not published

So I know there are different interpretations as precise @jcchoquet 

but we must think of the general spirit of fighting against the VAT cheating of traders at the level of their POS, for example http://www.fiscalonline.com/Logiciel-obligatoire-au-1er.html the lawyer
 Corinne Lecocq (our emblem anyway;) she explains :
 « Sont principalement visées les activités dispensées d’émission de factures, offrant une possibilité de dissimulation de recettes en espèces. C’est ce que l’administration vient de préciser afin d’éviter les incompréhensions des TPE/PME. Dans son communiqué de presse du 15 juin 2017, le Ministre précise en effet que seuls sont visés les logiciels de caisse. » donc une vision différente de vos interlocuteurs @jcchoquet 

And reminder, indulgences will be passed to the tax inspectors for the first year


For the certification, an attestation of the publisher can be sufficient but the ideal step would be a certification via one of the 2 accredited organization : LNE ou AFNOR
Patrick

**Comment by jcchoquet** [comment]
Hello @PatrickSUDOKEYS ,
actually according to the interlocutors, the versions are different, it's not easy.

**Comment by lap-odoo** [comment]
Hi All, 

@fgi-odoo and I would like to clarify some important things regarding the Certification for the French Accounting. 

1. The user will have the right to choose to install a module to certify the accounting. If the module is not installed, they won't be able to download the certification.

2. If this module is installed, it will apply on ALL accounting entries (not only the ones generated by the PoS). So all the accounting entries will be unalterable (and will receive an hash). 

3. Moreover, there are some additional requirements in the law. This will be implemented in the module (regarding the closing balance of the sales entries and the inalterability of the PoS orders). 

I hope this clarifies the situation

**Comment by PatrickSUDOKEYS** [comment]
Thank you Laura
It's clear. Odoo decides to go further than the law is top we will be ready for the future
Just a link to an LNE educational webinar : http://webinars.lne.fr/consommation/logiciels-caisse/index.asp?utm_source=signature&utm_medium=signature_webinar&utm_campaign=webinar_logiciels_caisse_fev17
Patrick

**Comment by sisalp** [comment]
2017-11-08 16:29 GMT+01:00 Laura Piraux <notifications@github.com>:

>
> @fgi-odoo and I would like to clarify some important things regarding the Certification for the French Accounting.

I guess that "French Accounting" is a shortcut, for LFP2016. "Sapin's
law" will not concern accounting by itself.

>
> If this module is installed, it will apply on ALL accounting entries (not only the ones generated by the PoS). So all the accounting entries will be unalterable (and will receive an hash).


So I understand here that tickets and payment information from the POS
are recorded in the accounting journals, and therefore, Sapin's law
constraints are moved from POS to accounting journals.

Thank you for this precision, if I got it correctly.

As a side consequence and just because it has been discussed here, I
don't think off-line POS is possible because my understanding is that
the POS will fail to pass the most important test of LNE procedure (n°
5 IIRC). Things are so complex that I hope I'm wrong.

**Comment by lap-odoo** [comment]
@sisalp : That's right, as the PoS is creating accounting entries, we ensure the protection of these accounting entries. But we will also ensure that the PoS Orders are unalterable.

**Comment by legalsylvain** [comment]
Hi @lap-odoo and @fgi-odoo. Thanks for this work with the community. It is very appreciated if Odoo SA can work with OCA on some technical points as l10n fr certification.

I realized some months ago a work to certify V8.0 Odoo PoS, regarding some points. The PR is here : https://github.com/OCA/l10n-france/pull/108
In a few words : 
-> it is a backport of the behaviour of the Odoo module l10n_fr_certification made by Odoo SA for the V9.0.
-> the code has been splited in 3 modules. **l10n_fr_certification_abstract** that provides a mixin model to certify any odoo model. (account.move, stock.picking, pos.order) = generate a new sha1, based on the data of an object, and the sha1 of the previous object.
-> a module **l10n_fr_certification_account** that make the same thing as the Odoo 9.0 module. the aim is to have a module that generate the same sha1 as the Odoo SA V9.0 module to have the possibility to switch from the OCA version to the Odoo version. (BTW I fixed a bug in the core here : https://github.com/odoo/odoo/pull/18175)
-> a module **l10n_fr_certification_pos** to certify pos.order as for account.move AND add extra PoS features. (possibility to block offline features, AND / OR possibility to write the sha1 on the pos bill). This part was not covered by the original work of Odoo SA. See @qdp-odoo https://github.com/odoo/odoo/pull/16935#issuecomment-301832280 remarks in May.

By the way, @qdp-odoo, did you took the time to check this PR ? (https://github.com/odoo/odoo/pull/16935#discussion_r126893261)

I'm available to work together on developpment. (sharing code, reviewing PR). Just let me know.
In all cases, it's important that OCA V8 algorithm renders the same result (for the sha1 computation) as the V9 / V10 Odoo SA modules. otherwise, it will not be possible for OCA 8.0 customers to migrate into a standard Odoo version. (EE / CE)

kind regards.

**Comment by fgi-odoo** [comment]
We will provide the PR link asap (on Odoo 9). We are finalizing it. In the meantime here is the draft of the related documentation, summarizing all the features: https://docs.google.com/document/d/1zAA_Qe2H7fCPvGbH_ztuMoNNzmT3xOmX__xF5Ugi2F8/edit?usp=sharing

**Comment by legalsylvain** [comment]
Thanks @fgi-odoo. I added some comments / questions in your docs. Please ping me when you have a PR ready to review. Note : what the name of the new module ? (fr certification for pos). i'd like to avoid collision with the OCA 8.0 PR.
regards.

**Comment by sisalp** [comment]
2017-11-09 15:37 GMT+01:00 fgi-odoo <notifications@github.com>:
>
> We will provide the PR link asap (on Odoo 9). We are finalizing it. In the meantime here is the draft of the related documentation, summarizing all the features: https://docs.google.com/document/d/1zAA_Qe2H7fCPvGbH_ztuMoNNzmT3xOmX__xF5Ugi2F8/edit?usp=sharing
>
Thank you for the information.
On the technical side, it is difficult to undertsand how is proven the
coverage of the requirements. I guess you do have this picture on your
side.
On the organizational side, certificate and probably attestation also,
requires some documentation about how the requirements are met and
also how the editor is organized to manage versioning of the software
and evolutions of the regulation. As Odoo is not (AFAIK) ISO 9001, it
may require some care.
For this, the end-user is not responsible, the attestation is
considered as the proof that Odoo can answer to this kind of inquiry.

Anyway, it's good to know that we can propose a solution to our users.

**Comment by sisalp** [comment]
2017-11-09 16:04 GMT+01:00 Sylvain LE GAL <notifications@github.com>:
 i'd like to avoid collision with the
> OCA 8.0 PR.

Can you update us about what has been achieved in the community
regarding this regulation ?

**Comment by fgi-odoo** [comment]
Here is the PR: https://github.com/odoo/odoo/pull/20581
Related doc (in draft, more technical details to be added): https://docs.google.com/document/d/1zAA_Qe2H7fCPvGbH_ztuMoNNzmT3xOmX__xF5Ugi2F8/edit?usp=sharing

We look forward to your feedback! Thanks!

**Comment by fgi-odoo** [comment]
Who got the chance to test the module? We would like to hear from you to finetune before launching it. Thanks in advance!

**Comment by jcchoquet** [comment]
I have a community version 9 but without POS, I can just test this module:
https://github.com/odoo/odoo/tree/35f299a0f899d48ff28e1b2d9e1a819231a8c2fb/addons/l10n_en_certification ?? or not?

**Comment by legalsylvain** [comment]
hi @fgi-odoo. I'll take a time to test your PR this friday. Thanks for asking the point of view of the community.
regards.

**Comment by flotho** [comment]
Hi @legalsylvain 

@gaelTorrecillas and I will block our friday to test this with you.
we could start early in the morning

**Comment by legalsylvain** [comment]
and reviewing #108 too ? ;-)

**Comment by robinshakty** [comment]
Hi all,
I can also have some free time to test the PR on friday. We have several customers with different settings (single, restaurants, multi-sessions syncing, ...).
I was in touch with @legalsylvain since a couple of days for a future project based on these new modules.
For i @fgi-odoo, do you imagine available modules for v10 for the end of the year ?
Do you have a git repository where I can download the modules in dev ?
Regards

**Comment by fgi-odoo** [comment]
@jcchoquet I don't have access to your link. Anyway you can test both modules with your v9 community version with just POS installed.

@robinshakty the plan is to forward port the modules to v10 and v11 as soon as we are done with the last refinements. To do so, we wait for the feedback of the community. Yes for sure this will be released for all the versions by the end of the year.

We are a bit short of time but thanks to your commitment, I hope we will come up with a final and completed solution within days. Thank you all!!

**Comment by flotho** [comment]
Hi @robinshakty , @legalsylvain , @fgi-odoo 

We're about to start the tests with @gaelTorrecillas. Could we do a summary of the environment and plan tests ? What is the appropriate process ? 
I propose to Test the  PR of @fgi-odoo https://github.com/odoo/odoo/pull/2058 first and then the #108 of @legalsylvain ?

Regarding the tests on the https://github.com/odoo/odoo/pull/20581, my interpretation is to proceed like this :
1. Check the hash generation of any account.move from account, POS etc... 
2. Check that it's not possible to modify invoices, payments, account.move once posted, deeply through the API
3. Check that automatic closing works correctly as proposed in @fgi-odoo  document https://docs.google.com/document/d/1zAA_Qe2H7fCPvGbH_ztuMoNNzmT3xOmX__xF5Ugi2F8/edit#heading=h.7e2xt3e5ild0
4. Test the process that identify which part of the move is modified 

@fgi-odoo, @legalsylvain maybe you could give us additionnal subject to test.


Regarding the test for @legalsylvain POS PR #108, If I understand well
1. Latest Odoo v8 pulled from Odoo
2. Latest code from the @legalsylvain PR : https://github.com/OCA/l10n-france/pull/108
3. Check that 3 modules are installed : l10n_fr_certification_abstract, l10n_fr_certification_account, l10n_fr_certification_pos
4. (The test of the PR of @fgi-odoo https://github.com/odoo/odoo/pull/20581 will be backported in the v8, once tested)
5. Check that the hash code appears on documents
6. Check option to prevent printing in offline mode, 

Anymore proposals will be appreciated.

regards

**Comment by robinshakty** [comment]
Hi all,
Firts of all, many thanks to the team for the job.

I just finished th first tests of these 2 modules.
1 . Install modules : ok, with no open pos session
2 . check inalterability before opening pos session : ok (message 'there is no order ...)
3 . Open pos session and creates several pos orders : ok
4. Check inalterability before closing pos session : ok 'Successful test ! The point of sale orders are guaranteed to be in their original and inalterable state 
From: Main/0004 (Receipt ref.: Commande 906069612) recorded on 17/11/2017 09:08:21
To: Main/0007 (Receipt ref.: Commande 906486460) recorded on 17/11/2017 09:14:55'
5. Close and post pos session : ok
6. Check inalterability of journal entries after closing pos session : ok 'Successful test ! The journal entries are guaranteed to be in their original and inalterable state
From: CSH4/2017/0002/1 (ref.: POS/2017/11/17/03)
To: FAC/2017/0004 (ref.: POS/2017/11/17/04)'

7.Check  the 3 Cumulative Grand Total : KO - they are not updated after the pos session closure.
I understood there is a cron to update them, btw I think these 3 Cumulative Grand Total have to be updated when one close and post the pos session.

In another way, I liked how @legalsylvain did the backport in v8 with a custom view to check the hash in a new pos order page and with printing the partial hash on the pos receipt.

Finally, to maintain an absolute inalterability, with managing backups and restores, I think it would be interesting to export and sign an xml file of the pos order in a path to define in the config file, just after saving the pos order (or invoice ...) in the database. It is just an idea...

**Comment by flotho** [comment]
Hi @robinshakty, 

maybe we could add those tests process in the PR https://github.com/OCA/l10n-france/pull/108 so that this issue won't be "ugly" to read.

**Comment by robinshakty** [comment]
If someone wants to do some tests, I give this link
http://odoo.alpimarket.com:52969
database : TST_FR_CERT_POS_V9
login : demo_cert
password : democert

I will check the hash with changing some data in sql or with export / import.
for @flotho , ok, how to move my comments ? (I am a beginner as a git user ...)
regards

**Comment by flotho** [comment]
@robinshakty , maybe copy paste first and then edit your comment,

The database name you proposed are all tagged v9, what test are you trying to do ? the #108 (fro v8 only) or the @fgi-odoo PR on v9?

**Comment by robinshakty** [comment]
@flotho , I tested  the @fgi-odoo PR on v9,
I already tested the @legalsylvain PR on v8 some weeks ago and what is done was well done.
regards

**Comment by flotho** [comment]
Ok, @robinshakty, thanks for your comment,
If you tested the #108, could you put some comment or result test on it so that @legalsylvain and the community could have some feedback.
Additionally, regarding testing this #108, what do you think about my tests proposal https://github.com/OCA/l10n-france/issues/105#issuecomment-345183078

**Comment by fgi-odoo** [comment]
Thanks @robinshakty for your feedback! Has anybody else tested it? Thanks!

**Comment by legalsylvain** [comment]
Hi @fgi-odoo, I already reviewed it directly on the PR https://github.com/odoo/odoo/pull/20581
regards.

**Comment by tkFontaine** [comment]
Odoo will give a module for the certification for Odoo community, but it means that **we have to obtain certification on our own**, right ?
Will you provide us all documents (or a part) for the certification ? One of organization (LNE) list all documents [here](https://www.lne.fr/sites/default/files/bloc-telecharger/referentiel-certification-systemes-caisse.pdf) at page 8.

**Comment by fgi-odoo** [comment]
@robinshakty thanks a lot for your suggestions. 

Why would it be better to process a sales closing at the closing of a pos session? You might close your POS several times a day or have several sessions running at the same time. It would considerably increase the number of sales closings. If it's only for control reasons, is it not acceptable to launch the cron manually? This will be explaned in the user doc.

Then regarding the xml file, this is adding a new layer of security (with another hash computation on the file I guess) that can hacked as well by a "man of the art". Also this extra process would slower the recording a pos order. The initial # computation makes it already a bit slower. So I'm not sure we should do that. A man of the art will always be able to change the hash, whatever you do to prevent it.

**Comment by fgi-odoo** [comment]
@tkFontaine yes you got it. Odoo will provide a user guide with some technical hints. Here is the draft: https://docs.google.com/document/d/1zAA_Qe2H7fCPvGbH_ztuMoNNzmT3xOmX__xF5Ugi2F8/edit?usp=sharing

**Comment by robinshakty** [comment]
Hi @fgi-odoo,
thank you for your responses and questions.
Why would it be better to process a sales closing at the closing of a pos session?
-> Because the sessions have to be closed every day and the software has to block new sales in an old session. You dit it well, and the law wrote that the user must use a system which has to calculate  daily grand totals, not when he think, or forget, or decide ...
It is why I suggested this possibility.

The idea of the hashed xml file is to have a hashed data out of the (backup of) the database, to maintain inalterability even after a restore of a database, with the possibility to sync with a digital safe (via rsync or ftp), and we would solve the problem of a 'man of the art' because hashed datas woul be secure by another 'real' and independant Certificate Authority.
I know that the law did not force this, but any ineltaribility with no independant Certificate Authority is 'relative', so is not 'ineltaribility'.

**Comment by fgi-odoo** [comment]
@robinshakty Ok, a user will not be able to resume a pos session after more than 24 hours. This will force them to process daily closings and so, to include pos sales into daily sales closings. There can be a gap of one day of course, if the sales closing occurs before the pos closing. But this can be easily figured out thanks to datetime fields on both objects. We don't see that as a problem.

Regarding the xml file I keep this nice suggestion to improve the module in master afterwards.

**Comment by fgi-odoo** [comment]
Here is our official statement: https://www.odoo.com/fr_FR/blog/notre-blog-5/post/obtenez-la-certification-anti-fraude-a-la-tva-avec-odoo-464.

The module will be available very soon.

**Comment by sisalp** [comment]
2017-12-13 15:15 GMT+01:00 fgi-odoo <notifications@github.com>:
> Here is our official statement:
> https://www.odoo.com/fr_FR/blog/notre-blog-5/post/obtenez-la-certification-anti-fraude-a-la-tva-avec-odoo-464.

Thank you, but care about the fact that Odoo is NOT certified, and in
this statement you must read attestation in place of certificate
sed s/certificate/attestation/g

>
> The module will be available very soon.
>
and will be interested in any update about your certification process.

**Comment by tkFontaine** [comment]
@fgi-odoo Thank you for your work.
I tested your module, and I can't Check _data inalterability_ on an existing database. I mean, I want to test your module on an database where there are already sales.
First, I clicked on _Data Inalterability Check_ (on POS), then I got the message 

> There isn't any order flagged for data inalterability yet for the company Crocali. This mechanism only runs for point of sale orders generated after the installation of the module France - Certification CGI 286 I-3 bis. - POS

Then I closed the session actually open, ~~and I reopened a new one, did some sales and closed it again. But I still get the same message . Is there a configuration to do ?~~ Is it a problem if there are already old sales before install modules ?

_Edit:_ I selected the wrong country.

But there still a question _Is it a problem if there are already old sales before install modules ?_

**Comment by fgi-odoo** [comment]
@sisalp We made sure to comply with every single requirement of the new law. So we are confident in delivering the certificate ourself. By law it is also forbidden to communicate anything about any certification process.

@tkFontaine all the old sales should be encrypted when you install the module.

**Comment by sisalp** [comment]
2017-12-18 10:00 GMT+01:00 fgi-odoo <notifications@github.com>:
> @sisalp We made sure to comply with every single requirement of the new law.
> So we are confident in delivering the certificate ourself.

I don't underestimate Odoo's self-confidence but words have their importance.
When you "certify yourself", you can deliver an "attestation"
When certification is done by an authority (there are two), you get a
certified product

Approximation is to be avoided at this step.

> By law it is also
> forbidden to communicate anything about any certification process.
>

Any reference to this law ? Certification process is made public by
certification authorities. I referenced the main one previously. Do
you mean "certification process results" instead ?


regards

**Comment by odony** [comment]
On 12/18/17 12:06 PM, sisalp wrote:
> Certification process is made public by
> certification authorities. I referenced the main one previously. Do
> you mean "certification process results" instead ?

http://www.nf525.com/index.php?option=com_content&view=article&id=95&Itemid=582
See §3.2.1.

**Comment by sisalp** [comment]
2017-12-18 12:33 GMT+01:00 Olivier Dony <notifications@github.com>:
> On 12/18/17 12:06 PM, sisalp wrote:
>> Certification process is made public by
>> certification authorities. I referenced the main one previously. Do
>> you mean "certification process results" instead ?
>
> http://www.nf525.com/index.php?option=com_content&view=article&id=95&Itemid=582
> See §3.2.1.
>

Yes, it confirms that you cannor claim to be certified before you
actually get certified by the authority.
This confirms you must say 'attestation' until then.

**Comment by tkFontaine** [comment]
> @fgi-odoo By law it is also forbidden to communicate anything about any certification process.

Will it includes all documents required for certification ? If yes, it means that we have to provide those documents by ourself ?

Last question, but not the least : Can someone tell me the difference between "**certificat**" and "**attestation individuelle**" ?

**Comment by Yakulu** [comment]
@tkFontaine there are detailed elements on BOFIP, around paragraph 350 http://bofip.impots.gouv.fr/bofip/10691-PGP

**Comment by fgi-odoo** [comment]
As discussed here a few months ago, printing the hash on the ticket is needed to avoid offline fraud (cfr https://github.com/odoo/odoo/pull/16935#issuecomment-301802227). We are working on such an improvement for the l10n_fr_pos_cert module. The ticket printing will therefore require to be online.

When it comes to restaurants, we need your opinion regarding pro forma printings. For now, you can take an order, print the pro forma and delete the order without registering the payment. This seems to go against the new legislation requiring to make inalterable any order update with + and - operations (items 20 of the FAQ https://www.economie.gouv.fr/files/files/directions_services/dgfip/controle_fiscal/actualites_reponses/logiciels_de_caisse.pdf).

As a solution, we think of recording all the printings in the backend (attached to the pos order) and prevent from deleting orders already printed in pro forma. That way, any update or deletion operation will be printed, saved and encrypted.

Do you agree on the solution? Thanks!

**Comment by legalsylvain** [comment]
Hi @fgi-odoo. Thanks a lot for sharing your insights !

> As discussed here a few months ago, printing the hash on the ticket is needed to avoid offline fraud (cfr odoo/odoo#16935 (comment)). We are working on such an improvement for the l10n_fr_pos_cert module. The ticket printing will therefore require to be online.

Regarding the print of the hash on the bill, I worked some monthes ago on that feature.
The description is available [here](https://github.com/grap/l10n-france/blob/b54f7f81b59d2d1a4bcdc39d1729f916a78a2e34/l10n_fr_certification_pos_offline/README.rst) 

- Could you [contact me](https://twitter.com/legalsylvain) to see if it's possible to include this work or part of this work into the Odoo Core ?
- I think it will be valuable to keep the "warning" feature for the PoS. Offline mode **should** be a possibility. Otherwise, Odoo PoS will not be a choice in France. We can not propose a PoS that doesn't work if there is a network problem.

![image](https://user-images.githubusercontent.com/3407482/41590649-cb9d88d0-73b7-11e8-8158-81940bb79340.png)

>  As a solution, we think of recording all the printings in the backend (attached to the pos order) and prevent from deleting orders already printed in pro forma. That way, any update or deletion operation will be printed, saved and encrypted.
> Do you agree on the solution? Thanks!

> prevent from deleting orders

+1 in the JS part. We could add a flag on order lines, that indicate that order line has been printed.

>  recording all the printings in the backend 

Is it necessary ? If the order lines can not be deleted or altered, once printed. The order will be validated. don't you think ? 


Other problem I see, unfortunately, in Bar and restaurant, sometimes, the waiter will give the "profarma", and the customer will go without pay. (This occures in France;-)). I think we should handle this case with an option to validate a pos order, but not paid. (something like with a Profit / Loss mechanism).

Kind regards.

**Comment by legalsylvain** [comment]
HI @fgi-odoo. Did you take a look on my last comment ?
kind regards.

CC @mart-e

**Comment by legalsylvain** [comment]
@fgi-odoo, @mart-e. No news about possible collaboration ?

**Comment by mart-e** [comment]
Hi,

As discussed by email, we would gladly collaborate on this :slightly_smiling_face:  
We do plan to add the signature. Having a warning if offline seems a good intermediate solution.
Let's work on this !

**Comment by rgeromegnace** [comment]
> The module will be available on community version of course. As long as Odoo provides the certificate, this will be only for Enterprise users. However we plan to get certified by an official certifier early 2018. If this really happens the certifier would distribute the certificate to any Odoo user.

Hello, I read all threads. What about the official certifier for Odoo Pos Community?

Regards,

**Comment by rgeromegnace** [comment]
Any answer @fgi-odoo ?

Regards

**Comment by fgi-odoo** [comment]
@rgeromegnace Modules used for the certification are available in the community version. But the certificate is only provided to Odoo Enterprise Users, as it implies a strong commitment from the editor (see https://www.odoo.com/fr_FR/blog/notre-blog-5/post/mise-en-conformite-de-votre-systeme-odoo-pour-la-nouvelle-legislation-anti-fraude-a-la-tva-464). If you run the community version, you should issue the certificate yourself or ask an integrator.

**Comment by rgeromegnace** [comment]
> 
> 
> @rgeromegnace Modules used for the certification are available in the community version. But the certificate is only provided to Odoo Enterprise Users, as it implies a strong commitment from the editor (see https://www.odoo.com/fr_FR/blog/notre-blog-5/post/mise-en-conformite-de-votre-systeme-odoo-pour-la-nouvelle-legislation-anti-fraude-a-la-tva-464). If you run the community version, you should issue the certificate yourself or ask an integrator.

Thanks for your reply even if it is late. I thought that there were partners or integrator that sold the module and that managed the certification.

Regards,
